### ISSUE TITLE
[Proj4] build failure - hardcoded paths

### ISSUE THREAD

**Issue by larshg** [microsoft/vcpkg#10913]
**Host Environment**
- OS: Windows
- Compiler: VS2019

**To Reproduce**

Create custom triplet file `x86-windows-rel.cmake` which build only release:

```
set(VCPKG_TARGET_ARCHITECTURE x86)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE dynamic)
set(VCPKG_BUILD_TYPE release)
```

`./vcpkg install proj4`

**Failure logs** 
Proj4 database need to install sqlite3[tool]:x86-windows first.

The path searched for is:
`${CURRENT_INSTALLED_DIR}/../x86-windows/tools`

The path where the exe resides is:
`installed\x86-windows-rel\tools`

**Additional context**
Add any other context about the problem here, such as what you have already tried to resolve the issue.

**Comment by LilyWangL** [comment]
Thanks for posting this issue. I created triplet file `x86-windows-rel.cmake` and installed proj4 successfully. Could you please provide the error log?

**Comment by larshg** [comment]
Ahh, I think I forgot to add [database] to the install command. It should be:
`./vcpkg install proj4[database]`

**Comment by larshg** [comment]
Will try post a log soon, but I'll just have to reproduce the error in my docker build. And figure out how to either copy the file or get it shown in powershell.

**Comment by larshg** [comment]
There aren't much more in the log really:

I think the checks here have a too fixed path:
https://github.com/microsoft/vcpkg/blob/bf338dc947348a1d9708acb4cda344ca94b69ead/ports/proj4/portfile.cmake#L25-L43

ie. the triplet part of the filepath, should be a variable with the actual triplet being used?

**Comment by JackBoosY** [comment]
The problem is that when we build `sqlite3:arm-windows`, `sqlite3.exe` will also be generated as arm type and can **only** be used in arm windows. When we use `x86` / `x64` windows to build `proj4[database]`, we must use `sqlite3.exe` to generate data files. At this time, directly using the arm type of `sqlite3.exe` will prompt the "execution of wrong platform".

**Comment by larshg** [comment]
I'm compiling on x86, not on arm, so what does this has to do with anything?

The check here: 
`if (EXISTS ${CURRENT_INSTALLED_DIR}/../x86-windows/tools/sqlite3.exe)`

tries to find sqlite3.exe in the folder path with the triplet part as foldername `x86-windows` but if I use a new triplet file called `x86-windows-rel` it fails to locate it, even though it is build.

I guess what needs to be changed is to use a path that is generated from `VCPKG_TRIPLET_NAME` or something similiar, so the full path will be:
`if (EXISTS ${CURRENT_INSTALLED_DIR}/../${VCPKG_TRIPLET_NAME}/tools/sqlite3.exe)`

which will make it look in the correct folder.

**Comment by JackBoosY** [comment]
@larshg Since the database file that proj4 needs to generate is generated by using `sqilte3.exe`, and the `sqlite3.exe` generated under triplet `arm` / `arm64` cannot be used in non `arm` / `arm64`, we must first generate `sqlite3.exe` for `x86` / `x64` to build it normally.

![image](https://user-images.githubusercontent.com/47264268/80296392-56963d00-87ad-11ea-946f-83e20934b4b8.png)

**Comment by rossctaylor** [comment]
I am having the same issue. I was able to fix it by adding the following:
```
        if (EXISTS ${CURRENT_INSTALLED_DIR}/../${TARGET_TRIPLET}/tools/sqlite3.exe)
            set(SQLITE3_BIN_PATH ${CURRENT_INSTALLED_DIR}/../${TARGET_TRIPLET}/tools)
```
It isn't entirely clear to me what the logic is doing so I'm not sure this is a global fix. I understand that an exe built for arm won't work on x86 but I don't understand how the current logic is intended to address that issue. It checks if target is windows but not whether it is arm. If target and host are both windows and x64 (my case), it should just use the TARGET_TRIPLET. Would something like below be more appropriate (sorry, not tested)?
```
if ("database" IN_LIST FEATURES)
    if (CMAKE_HOST_WIN32)
        set(BIN_SUFFIX .exe)
        if ((VCPKG_TARGET_ARCHITECTURE STREQUAL "x86" OR (VCPKG_TARGET_ARCHITECTURE STREQUAL "x64" AND (NOT CMAKE_HOST_SYSTEM_PROCESSOR OR CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL "x86_64")) AND (EXISTS ${CURRENT_INSTALLED_DIR}/../${TARGET_TRIPLET}/tools/sqlite3.exe))
            set(SQLITE3_BIN_PATH ${CURRENT_INSTALLED_DIR}/../${TARGET_TRIPLET}/tools)
        elseif (EXISTS ${CURRENT_INSTALLED_DIR}/../x86-windows/tools/sqlite3.exe)
            set(SQLITE3_BIN_PATH ${CURRENT_INSTALLED_DIR}/../x86-windows/tools)
        elseif (EXISTS ${CURRENT_INSTALLED_DIR}/../x86-windows-static/tools/sqlite3.exe)
            set(SQLITE3_BIN_PATH ${CURRENT_INSTALLED_DIR}/../x86-windows-static/tools)
        elseif (EXISTS ${CURRENT_INSTALLED_DIR}/../x64-windows/tools/sqlite3.exe AND (NOT CMAKE_HOST_SYSTEM_PROCESSOR OR CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL "x86_64"))
            set(SQLITE3_BIN_PATH ${CURRENT_INSTALLED_DIR}/../x64-windows/tools)
        elseif (EXISTS ${CURRENT_INSTALLED_DIR}/../x64-windows-static/tools/sqlite3.exe AND (NOT CMAKE_HOST_SYSTEM_PROCESSOR OR CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL "x86_64"))
            set(SQLITE3_BIN_PATH ${CURRENT_INSTALLED_DIR}/../x64-windows-static/tools)
        else()
            message(FATAL_ERROR "Proj4 database need to install sqlite3[tool]:x86-windows first.")
        endif()
    else()
        set(BIN_SUFFIX)
        set(SQLITE3_BIN_PATH ${CURRENT_INSTALLED_DIR}/tools)
    endif()
endif()
```

**Comment by JackBoosY** [comment]
@rossctaylor There are two scenarios for building arm programs: host is non-arm and host is arm. 
So we need to judge not only local_host but also target_host.

**Comment by rossctaylor** [comment]
That's what I tried to do with the second proposal. That first long if statement attempts to check whether target and host match (both x86 or both x64). If so, it uses the target triplet executable. If not, it falls back to the logic that was already there to just search for an existing install of x86 or x64. 

The only other change I suggested was changing the second level `if (VCPKG_TARGET_IS_WINDOWS)` to `if (CMAKE_HOST_WIN32)` because we are searching for exe on the host so it makes more sense (to me) to check whether the host is windows rather than the target.
